### PR TITLE
api: wasm extension

### PR DESCRIPTION
**Pull request by zhaohuabing**
This PR proposes an API for supporting Wasm plugins:

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: EnvoyExtensionPolicy
metadata:
  name: wasm-extensions
spec:
  targetRef:
    kind: HTTPRoute
    name: backend
  wasm:
    - name: my_plugin_1
      code:
        image:  # the oci image that contains wasm code
          url: <oci-registry>.foo.bar.com/wasm-to-oci:v1 
          pullSecret: wasmPullSecret
        # http: https://foo.bar.com/example.wasm  # the HTTP URL containing the wasm code.
      sha256: xxxxxxx # SHA256 checksum that will be used to verify the wasm code.
      failOpen: true
      config:
        foo:
           foo1: value1
           foo2: value2
        bar: value3
      priority: 450   
    - name: my_plugin_2
      ...
```

### CODE REVIEW

File: api/v1alpha1/wasmextensionpolicy_types.go
Comment (Xunzhuo) [inline api/v1alpha1/wasmextensionpolicy_types.go:58]: Should we add `fail_open` and `capability_restriction_config` on WasmExtension ?

**Comment by guydc** [comment]
What persona will be using this API? IMHO it's a bit dangerous to let application owners inject code into envoy, even if executed in isolation. If we see this as a policy for Gateway admins, should we support attachment to routes?

**Comment by zhaohuabing** [comment]
My apologies,  I haven't noticed the existing discussion and the draft PR and didn't mean to add a competing PR.  :-)

Before proposing the API, I considered two approaches to support data plane extensions (WASM, Lua, Golang, External Process, etc.
* A fat EnvoyExtensionPolicy containing all the extensions.
* A set of EnvoyXExtensionPolicy, each supporting a specific extension.

The main problem of the first approach is that it's hard to support multiple extensions of the same type, e.g. users may need to define multiple WASM filters for different purposes for a single HTTPRoute.

File: api/v1alpha1/wasmextensionpolicy_types.go
Comment (zhaohuabing) [reply api/v1alpha1/wasmextensionpolicy_types.go:58]: Yes, they can be added later. I just want to make the first API as simple as possible.

**Comment by guydc** [comment]
> My apologies @guydc , I haven't noticed the existing discussion and the draft PR and didn't mean to add a competing PR. :-)

All good, I'm very happy that this topic is getting more attention and I hope that we can make this happen for v1.1.0. 

> Before proposing the API, I considered two approaches to support data plane extensions (WASM, Lua, Golang, External Process, etc.)

I agree that different extensions do have different structures, interfaces, concerns and use patterns. Some support per-route config while other don't, some are more modular and may repeat multiple time in the FC, while others are external and more likely to only be called-out once in the filter chain. 

I'm very much open to different extension types being supported for different use cases if other maintainers also support this approach.

**Comment by zhaohuabing** [comment]
> I agree that different extensions do have different structures, interfaces, concerns and use patterns. Some support per-route config while other don't, some are more modular and may repeat multiple time in the FC, while others are external and more likely to only be called-out once in the filter chain.

> I'm very much open to different extension types being supported for different use cases if other maintainers also support this approach.

Let's bring this to the EG meeting and discuss further.  Even we option to have multiple extension types, these types do need to address the common concerns raised in https://github.com/envoyproxy/gateway/pull/2570 to ensure cohensive approach.

* Users may need to control the logical phase of filter execution (e.g. before auth, before rate limit, ... )
* Users may need to control the order of execution of extension filters amongst themselves (within a specific phase)

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [inline api/v1alpha1/wasm_types.go:36]: Is there a better way to represent protobuf any.Any in the API?

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [inline api/v1alpha1/wasm_types.go:37]: This depends om https://github.com/envoyproxy/gateway/pull/2993

Comment (arkodg) [reply api/v1alpha1/wasm_types.go:42]: here is my suggestion for prioritization of dealing with filter ordering to so we move faster in the API reviews
cc @guydc 
1. lets find a default home (order) for `ext_proc` and `wasm` (maybe right after `ext_auth`)
2. lets create per proxy ordering (which you have already done with #2993 (love it)) once we get use cases in
3. lets then consider creating per wasm location override once we get use cases in

Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go]: @arkodg We may need to address ordering first because different WASM extensions should be in different locations in the chain.

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [inline api/v1alpha1/wasm_types.go:37]: Envoy supports struct, string, and bytes. Follow Istio practice to restrict input as a JSON string for a better UI:

* We can avoid media disruption in the K8s resources by just specifying the config as a YAML object. JSON is a subset of YAML, so it is very easy to convert to it from YAML.
* JSON has parsers in almost every language (as opposed to e.g. YAML), and certainly the ones that already compile to Wasm. It is also compatible with protobuf, should we want to use that in some of our SDKs.
* Developers don’t have to deal with parsing configuration. The language SDKs can use optimized code to unmarshall the JSON configuration and make the values easily accessible to developers in a language-specific way (this could e.g. be a map[string]interface{} in golang).
* Using a structured format for configuration allows for configuration injection and templating


https://docs.google.com/document/d/15GQnhOpLfvgyOZ8VDF5KD-q2cjXEMLdTQMm0YPswIu0/edit#heading=h.6zqg8n4aice5

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [inline api/v1alpha1/wasm_types.go:83]: OIC image is not supported by Envoy. So EG needs to download the wasm module from the OIC registry and mount it to the Envoy container.

Comment (zirain) [reply api/v1alpha1/wasm_types.go:83]: there's no agent, we need to do this on controller?

Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:83]: IMO, EG doesn't need an agent like Istio does, since the Control Plane and Data Plane are in the same cluster.

Comment (zirain) [reply api/v1alpha1/wasm_types.go:83]: envoy support http, EG can host an http endpoint?

**Comment by arkodg** [comment]
hey can the PR description be used to reflect the newer API ?

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [reply api/v1alpha1/wasm_types.go:83]: a lot of the users already have their existing WASM module in a OCI registry so we should support this use case

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:13]: can we autogenerate `name`, `vm_id` and `root_id` ? and leave it out of the first iteration ?

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:12]: maybe some comments here around defaults, such as only `V8` type and `HTTP` filter based WASM modules are supported today

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:13]: We can generate the `name`, but the auto-generated name might not be as meaningful as the user-provided name. 

For `vm_id` and `root_id`, we might want to expose them to the API to allows users to share VM or RootContext.

**Comment by zhaohuabing** [comment]
> hey can the PR description be used to reflect the newer API ?

Glad to see you back! @arkodg 

The description was modified.

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [inline api/v1alpha1/wasm_types.go:69]: Should we support ConfigMap? Considering the size limit of CM(1M), it might not be very useful.

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:62]: should we start with only supporting 2 types - http and image ?
something like
```
type: "Image"
image:
  url:
  pullPolicy:
  pullSecret:
```

Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:62]: Yah, I also think CM is not so useful in production.

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:61]: sidenote /non blocker - can code be refreshed w/o restarting Envoy ?

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [reply api/v1alpha1/wasm_types.go]: yeah we can tackle before v1.1 , but it doesnt need to block this PR

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:61]: I believe so. WASM filter is an HTTP filter in the HCM filter chain and the code is loaded on the fly.

Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:61]: I think I need to come up with a simple design doc before implementation.

Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:65]: > sidenote /non blocker - can code be refreshed w/o restarting Envoy ?

Seems that I misunderstood your comment :-) 

I guess we could change the sha256 to update the wasm code. I haven't verified that yet.

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:17]: @arkodg Envoy requires `name` to be unique within a VM. Since we have decided to leave vm_id out, EG will use a separate VM for each wasm extension, the name doesn't have to be unique in EnvoyExtensionPolicy.

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:80]: lets rm SHA for now, until we decide how content pull will work ?

File: api/v1alpha1/wasm_types.go
Comment (zirain) [inline api/v1alpha1/wasm_types.go]: should remove it instead of comment.

Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go]: I leave this because it's likely to be used when implementing wasm extension.  

Of course, I'll remove it if I find out it's not necessary when implementing it.

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:90]: ratelimit ? :)

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:116]: lets rm `pullPolicy` for now as well since its a k8s specific knob thats living in a generic API

File: api/v1alpha1/wasm_types.go
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:110]: should this be pullSecretRef ?

File: api/v1alpha1/wasm_types.go
Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:116]: We probably can't reuse the k8s PullPolicy ?

```
const (
	// PullAlways means that kubelet always attempts to pull the latest image. Container will fail If the pull fails.
	PullAlways PullPolicy = "Always"
	// PullNever means that kubelet never pulls an image, but only uses a local image. Container will fail if the image isn't present
	PullNever PullPolicy = "Never"
	// PullIfNotPresent means that kubelet pulls if the image isn't present on disk. Container will fail if the image isn't present and the pull fails.
	PullIfNotPresent PullPolicy = "IfNotPresent"
)
```

Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go]: Remove it for now.

**Review by arkodg — APPROVED** [review]
thanks !

**Review by zirain — APPROVED** [review]

**Comment by arkodg** [comment]
hey @zhaohuabing lets start off with `HTTP` support, and temporarily pause on the design for the `Image` feature based on the discussions from https://github.com/envoyproxy/envoy/issues/33212

**Pull request merged**
